### ISSUE TITLE
Rewrite in Rust

### ISSUE THREAD

**Issue by liamdiprose** [elixir-sqlite/exqlite#245]
Hello, its me, "that guy".

I've been trying to debug a segfault coming from the beam and its not pretty.

I saw you recommend Rustler in another issue. Did you want to transition this library to Rust?

I've also noticed interest in support for other databases that support the SQLite3 interface, i.e. DuckDB and mvsqlite.

- We could turn  https://docs.rs/sqlite/latest/sqlite/struct.Connection.html into a Trait, then implement it for each database using either their Rust API, or their unsafe-C sqlite-wrapper[^1].
- Or, perhaps, there a more general Ecto trait can be made, meaning Rust database-libraries only need to implement what the Ecto wrapper requires.

Since you've actually written the Ecto bindings, you would have much clearer idea of what is appropriate. I'm happy to answer any questions about Rust, although I'm sure there's plenty of us lurking around here.

It would good to keep Explorer in mind too. (https://github.com/elixir-nx/explorer/issues/55)

[^1]: I don't mind debugging a segfault, its just much much easier in Rust than the beam.

**Comment by warmwaffles** [comment]
I do not want to transition this library to rust unfortunately. I don't know it well enough and honestly do not have the time to explore it. I've tried and failed multiple times to learn Rust.

If others want to fork this and provide their own Rust implementations, they are more than welcome to do so. I made `ecto_sqlite3` as flexible as possible to allow for a swap out for a different adapter to an sqlite database.
